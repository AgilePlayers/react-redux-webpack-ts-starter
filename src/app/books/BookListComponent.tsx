import React from 'react';
import BookItemComponent from './components/bookItem/BookItemComponent';
import { Book } from './_duck/types';
import './bookList.scss';

export interface BookListComponentProps {
  isLoading: boolean;
  books: Book[];
  error: Error | null;
  getData(): void;
}

class BookListComponent extends React.Component<BookListComponentProps> {
  componentDidMount() {
    this.props.getData();
  }

  renderBookItems() {
    return this.props.books.map((book: Book) => {
      const { id, volumeInfo } = book;
      return <BookItemComponent info={volumeInfo} key={id} />;
    });
  }

  render() {
    if (this.props.isLoading) {
      return (
        <>
          <h1>Books</h1>
          <div className="book-loader">Data is loading...</div>
          <div className="book-loader-spinner" />
        </>
      );
    }
    return (
      <>
        <h1>Books</h1>
        <ul className="book-list">{this.renderBookItems()}</ul>
      </>
    );
  }
}

export default BookListComponent;
