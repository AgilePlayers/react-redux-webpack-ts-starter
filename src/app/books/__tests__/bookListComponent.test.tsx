import React from 'react';
import { shallow, mount, ShallowWrapper, ReactWrapper } from 'enzyme';
import BookListComponent, {
  BookListComponentProps,
} from '../BookListComponent';

describe('books / BookListComponent', () => {
  let component: ShallowWrapper | ReactWrapper;
  let props: BookListComponentProps;

  beforeEach(() => {
    props = {
      isLoading: false,
      books: [
        {
          id: '1asv',
          volumeInfo: {
            title: 'Title',
            description: 'Description',
            imageLinks: {
              smallThumbnail: 'thumbnail',
            },
          },
        },
      ],
      error: null,
      getData: jest.fn(),
    };
  });

  test('should render', () => {
    component = shallow(<BookListComponent {...props} />);
    expect(component.length).toEqual(1);
  });

  test('should contains element with div with "counter" class', () => {
    component = shallow(<BookListComponent {...props} />);
    expect(component.find('ul.book-list').length).toEqual(1);
  });

  test('should render loader if isLoading prop equals true', () => {
    props = { ...props, isLoading: true };
    component = shallow(<BookListComponent {...props} />);
    expect(component.find('div.book-loader').length).toEqual(1);
    expect(component.find('div.book-loader').text()).toEqual(
      'Data is loading...',
    );
  });

  test('should componentDidMount be called', () => {
    const componentDidMountSpy = jest.spyOn(
      BookListComponent.prototype,
      'componentDidMount',
    );
    component = mount(<BookListComponent {...props} />);
    expect(componentDidMountSpy).toHaveBeenCalledTimes(1);
    componentDidMountSpy.mockClear();
  });

  test('should renderBookItems be called', () => {
    const renderBookItemsSpy = jest.spyOn(
      BookListComponent.prototype,
      'renderBookItems',
    );
    component = mount(<BookListComponent {...props} />);
    expect(renderBookItemsSpy).toHaveBeenCalledTimes(1);
    renderBookItemsSpy.mockClear();
  });

  test('should match snapshot', () => {
    component = shallow(<BookListComponent {...props} />);
    expect(component).toMatchSnapshot();
  });
});
