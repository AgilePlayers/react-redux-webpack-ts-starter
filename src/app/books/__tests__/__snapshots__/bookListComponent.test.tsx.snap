// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`books / BookListComponent should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BookListComponent
    books={
      Array [
        Object {
          "id": "1asv",
          "volumeInfo": Object {
            "description": "Description",
            "imageLinks": Object {
              "smallThumbnail": "thumbnail",
            },
            "title": "Title",
          },
        },
      ]
    }
    error={null}
    getData={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    isLoading={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <h1>
          Books
        </h1>,
        <ul
          className="book-list"
        >
          <BookItemComponent
            info={
              Object {
                "description": "Description",
                "imageLinks": Object {
                  "smallThumbnail": "thumbnail",
                },
                "title": "Title",
              }
            }
          />
        </ul>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Books",
        },
        "ref": null,
        "rendered": "Books",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <BookItemComponent
              info={
                Object {
                  "description": "Description",
                  "imageLinks": Object {
                    "smallThumbnail": "thumbnail",
                  },
                  "title": "Title",
                }
              }
            />,
          ],
          "className": "book-list",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "1asv",
            "nodeType": "function",
            "props": Object {
              "info": Object {
                "description": "Description",
                "imageLinks": Object {
                  "smallThumbnail": "thumbnail",
                },
                "title": "Title",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "ul",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <h1>
            Books
          </h1>,
          <ul
            className="book-list"
          >
            <BookItemComponent
              info={
                Object {
                  "description": "Description",
                  "imageLinks": Object {
                    "smallThumbnail": "thumbnail",
                  },
                  "title": "Title",
                }
              }
            />
          </ul>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Books",
          },
          "ref": null,
          "rendered": "Books",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <BookItemComponent
                info={
                  Object {
                    "description": "Description",
                    "imageLinks": Object {
                      "smallThumbnail": "thumbnail",
                    },
                    "title": "Title",
                  }
                }
              />,
            ],
            "className": "book-list",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "1asv",
              "nodeType": "function",
              "props": Object {
                "info": Object {
                  "description": "Description",
                  "imageLinks": Object {
                    "smallThumbnail": "thumbnail",
                  },
                  "title": "Title",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "ul",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
